In [1]:
# import libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns

In [2]:
# Suppress Warnings for clean notebook
import warnings
warnings.filterwarnings('ignore')

In [4]:
# read dataset
dataset = pd.read_csv('Melbourne_housing_FULL.csv')
dataset.head(3)

Suburb          Address  Rooms Type      Price Method SellerG  \
0  Abbotsford    68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford     85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford  25 Bloomburg St      2    h  1035000.0      S  Biggin   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/09/2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   

   YearBuilt         CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN  Yarra City Council  -37.8014    144.9958  Northern Metropolitan   
1        NaN  Yarra City Council  -37.7996    144.9984  Northern Metropolitan   
2     1900.0  Yarra City Council  -37.8079    144.9934  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  

[3 rows x 21 columns]

In [5]:
dataset.nunique()

Suburb             351
Address          34009
Rooms               12
Type                 3
Price             2871
Method               9
SellerG            388
Date                78
Distance           215
Postcode           211
Bedroom2            15
Bathroom            11
Car                 15
Landsize          1684
BuildingArea       740
YearBuilt          160
CouncilArea         33
Lattitude        13402
Longtitude       14524
Regionname           8
Propertycount      342
dtype: int64

In [6]:
dataset.shape

(34857, 21)

In [7]:
# let's use limited columns which makes more sense for serving our purpose
cols_to_use = ['Suburb', 'Rooms', 'Type', 'Method', 'SellerG', 'Regionname', 'Propertycount', 
               'Distance', 'CouncilArea', 'Bedroom2', 'Bathroom', 'Car', 'Landsize', 'BuildingArea', 'Price']
dataset = dataset[cols_to_use]

In [9]:
dataset.head(3)

Suburb  Rooms Type Method SellerG             Regionname  \
0  Abbotsford      2    h     SS  Jellis  Northern Metropolitan   
1  Abbotsford      2    h      S  Biggin  Northern Metropolitan   
2  Abbotsford      2    h      S  Biggin  Northern Metropolitan   

   Propertycount  Distance         CouncilArea  Bedroom2  Bathroom  Car  \
0         4019.0       2.5  Yarra City Council       2.0       1.0  1.0   
1         4019.0       2.5  Yarra City Council       2.0       1.0  1.0   
2         4019.0       2.5  Yarra City Council       2.0       1.0  0.0   

   Landsize  BuildingArea      Price  
0     126.0           NaN        NaN  
1     202.0           NaN  1480000.0  
2     156.0          79.0  1035000.0

In [10]:
dataset.isna().sum()

Suburb               0
Rooms                0
Type                 0
Method               0
SellerG              0
Regionname           3
Propertycount        3
Distance             1
CouncilArea          3
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21115
Price             7610
dtype: int64

In [11]:
# Some feature's missing values can be treated as zero (another class for NA values or absence of that feature)
# like 0 for Propertycount, Bedroom2 will refer to other class of NA values
# like 0 for Car feature will mean that there's no car parking feature with house
cols_to_fill_zero = ['Propertycount', 'Distance', 'Bedroom2', 'Bathroom', 'Car']
dataset[cols_to_fill_zero] = dataset[cols_to_fill_zero].fillna(0)

# other continuous features can be imputed with mean for faster results since our focus is on Reducing overfitting
# using Lasso and Ridge Regression
dataset['Landsize'] = dataset['Landsize'].fillna(dataset.Landsize.mean())
dataset['BuildingArea'] = dataset['BuildingArea'].fillna(dataset.BuildingArea.mean())

In [12]:
dataset.dropna(inplace=True)

In [13]:
dataset.shape

(27244, 15)

In [14]:
dataset = pd.get_dummies(dataset, drop_first=True)

In [15]:
dataset.head(3)

Rooms  Propertycount  Distance  Bedroom2  Bathroom  Car  Landsize  \
1      2         4019.0       2.5       2.0       1.0  1.0     202.0   
2      2         4019.0       2.5       2.0       1.0  0.0     156.0   
4      3         4019.0       2.5       3.0       2.0  0.0     134.0   

   BuildingArea      Price  Suburb_Aberfeldie  ...  \
1      160.2564  1480000.0              False  ...   
2       79.0000  1035000.0              False  ...   
4      150.0000  1465000.0              False  ...   

   CouncilArea_Moorabool Shire Council  CouncilArea_Moreland City Council  \
1                                False                              False   
2                                False                              False   
4                                False                              False   

   CouncilArea_Nillumbik Shire Council  CouncilArea_Port Phillip City Council  \
1                                False                                  False   
2                                False                                  False   
4                                False                                  False   

   CouncilArea_Stonnington City Council  CouncilArea_Whitehorse City Council  \
1                                 False                                False   
2                                 False                                False   
4                                 False                                False   

   CouncilArea_Whittlesea City Council  CouncilArea_Wyndham City Council  \
1                                False                             False   
2                                False                             False   
4                                False                             False   

   CouncilArea_Yarra City Council  CouncilArea_Yarra Ranges Shire Council  
1                            True                                   False  
2                            True                                   False  
4                            True                                   False  

[3 rows x 745 columns]

In [16]:
X = dataset.drop('Price', axis=1)
y = dataset['Price']

In [17]:
from sklearn.model_selection import train_test_split
train_X, test_X, train_y, test_y = train_test_split(X, y, test_size=0.3, random_state=2)

In [18]:
from sklearn.linear_model import LinearRegression
reg = LinearRegression().fit(train_X, train_y)

In [19]:
reg.score(test_X, test_y)

0.1385368316156511

In [20]:
reg.score(train_X, train_y)

0.6827792395792723

###### Here training score is 68% but test score is 13.85% which is very low

##### Using Lasso (L1 Regularized) Regression Model

In [21]:
from sklearn import linear_model
lasso_reg = linear_model.Lasso(alpha=50, max_iter=100, tol=0.1)
lasso_reg.fit(train_X, train_y)

Lasso(alpha=50, max_iter=100, tol=0.1)

In [22]:
lasso_reg.score(test_X, test_y)

0.6636111369404489

In [23]:
lasso_reg.score(train_X, train_y)

0.6766985624766824

##### Using Ridge (L2 Regularized) Regression Model

In [25]:
from sklearn.linear_model import Ridge
ridge_reg= Ridge(alpha=50, max_iter=100, tol=0.1)
ridge_reg.fit(train_X, train_y)

Ridge(alpha=50, max_iter=100, tol=0.1)

In [27]:
ridge_reg.score(test_X, test_y)

0.6670848945194956

In [28]:
ridge_reg.score(train_X, train_y)

0.6622376739684328

#### We see that Lasso and Ridge Regularizations prove to be beneficial when our Simple Linear Regression Model overfits. These results may not be that contrast but significant in most cases.Also that L1 & L2 Regularizations are used in Neural Networks too